tests/address-utils/test_address_utils.py

```diff
@@ -23,6 +23,8 @@
         ({}, False, False, False),
         ("function", False, False, False),
         (b"function", False, False, False),
+        # This has a weird upper case letter but should it be recognized as valid address?
+        ("0x6a0a0fc761c612C340a0e98d33b37a75e5268472", True, True, False),
```

Comment:
@carver @pipermerriam I ran into this weird case today. `is_address` returns `False` for this (hence the test fails) but I wonder if it should return `True` instead. It is a valid address but it you look closely it contains an upper case `C` but is **NOT** a valid checksum address. So what happens is that `is_address` looks at it, sees that it contains mixed casing and then returns `is_checksum_address(val)` which will return `False`.

My personal compass says this is wrong and despite the weird casing `is_address` should return `True` for this address. What do you think?

Reply:
For context, this is the kind of change that would require a major version bump. It's not a bug, in the sense that it's working as designed. (So now we discuss whether it's a good design)

Right now, one way to use the API is something like:
```py
if not is_address(random_input):
  raise ValidationError(f"{random_input} is not a valid address")
```

If we change it, it becomes something like:
```py
if not is_address(random_input):
  raise ValidationError(f"{random_input} is not a valid address")
elif is_checksum_formatted_address(random_input) and not is_checksum address(random_input):
  raise ValidationError(f"{random_input} is an invalid checksum")
```

It would be very easy to forget to do that second check. There are other cases where a typo in the address will cause it to return `False` for `is_address()`, like accidentally adding an extra character at the end, or dropping the leading 0 (but leaving the `x`). We can't guess at all the things that make it almost a valid address, but not quite, so it makes sense to me to only return true for a really valid address, and specifically to return `False` if the checksum is invalid.

Reply:
Ah ok, I wasn't sure if this is by design. I mean, I can certainly send money to it and it will arrive. Also etherscan finds it when I enter it in their search field.

![image](https://user-images.githubusercontent.com/521109/83235511-4699c080-a192-11ea-9149-acf151986883.png)

That said, MetaMask will reject it as invalid.

![image](https://user-images.githubusercontent.com/521109/83235555-5e714480-a192-11ea-8c30-cf38a0239abc.png)

I mean another idea would be to introduce an optional `ignore_casing` parameter that defaults to `False` which would preserve the current behavior but could be call `lower()` on the input if flipped to `True` but I don't really need that change. I just wasn't sure if the current behavior is actually intended. But seeing that MetaMask and Mycrypto handle it the same gives me some confidence that this is in fact intended behavior :)

Reply:
I think I'm in support of keeping it as-is but I'll say that we do have four address types that we recognize:

- hex: any hex string of length 40 (ignoring 0x prefix)
- canonical: length-20 byte string
- normalized: any hex string, all lower case, length 40 (ignoring 0x prefix)
- checksummed: EIP55 checksummed hex address

`is_hex_address` would indeed return `True` for the offending address here.